### ISSUE TITLE
How do you use gradient checkpointing if you are updating gradients with an optimizer?

### ISSUE THREAD

**Issue by ryx2** [openai/gradient-checkpointing#4]

**Comment by yaoli** [comment]
I think the following should do the trick. The reason is that ``Optimizers.compute_gradients()`` calls ``gradients.gradients()`` instead of ``tf.gradients()`` internally.
```
import memory_saving_gradients
from tensorflow.python.ops import gradients
# monkey patch tf.gradients to point to our custom version, with automatic checkpoint selection                        
def gradients_memory(ys, xs, grad_ys=None, **kwargs):
    return memory_saving_gradients.gradients(ys, xs, grad_ys, checkpoints='memory', **kwargs)
gradients.__dict__["gradients"] = gradients_memory 
```

**Comment by tano297** [comment]
Normally, you can replace the optimizer.minimize() with something like this:

```py
grads = tf.gradients(loss, tf.trainable_variables())
grads_and_vars = list(zip(grads, tf.trainable_variables()))
train_op = optimizer.apply_gradients(grads_and_vars)
```
Changing tf.gradients to point to the desired gradients function provided by this repo should do the trick.
